### PR TITLE
Correctly reset metric objects in self.log

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/7052
Since commit https://github.com/PyTorchLightning/metrics/commit/19b77cc2c290e811ae1c1ea44125776414b4e431 in torchmetrics, the integration with lightning has been broken if the metric object have been tried to be logged directly. This is because the lightning logging implicit depends on the `self._computed` variable *not* being reset whenever `reset` is called. 

This PR tries to solve this by making sure that `reset` of metric objects is only called once at the very end. Please note that I am not so familiar with the logger connection of the code base.

As an alternative we revert the changes in torchmetrics. However, it does make some sense that the `self._computed` variable should be reset when `reset` is called.

cc: @ethanwharris 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -587,6 +578,14 @@ def get_non_metrics_keys(self):
         """
         return [k for k, v in self.items() if not isinstance(v, Metric)]
 
+    def reset(self):
```
Comment (ananthsub) [inline pytorch_lightning/core/step_result.py:581]:
```suggestion
    def reset(self) -> None:
```

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 2]
```diff
@@ -233,6 +233,18 @@ def auto_reduce_results_on_epoch_end(self) -> None:
 
         self.has_reduced = True
 
+    def reset(self):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:236]:
```suggestion
    def reset(self) -> None:
```

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 3]
```diff
@@ -233,6 +233,18 @@ def auto_reduce_results_on_epoch_end(self) -> None:
...
+        """
+        Call at the end of epoch to reset Result objects
+        """
+        for dl_idx in range(self.num_dataloaders):
+            epoch_metrics = self._internals[dl_idx] if not self.has_reduced else self._internals_reduced[dl_idx]
+            if self._internal_type == ResultStoreType.INSIDE_BATCH_TRAIN_LOOP:
+                for opt_idx in list(epoch_metrics):
+                    epoch_metrics[opt_idx].reset()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:244]: could you explain this check?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 4]
```diff
@@ -443,6 +456,8 @@ def get_forked_metrics(self) -> Dict:
         return self.run_epoch_by_func_name("get_forked_metrics")
 
     def reset(self):
+        for k, value in self._internals.items():
+            value.reset()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:460]:
```suggestion
    def reset(self) -> None:
        for k, value in self._internals.items():
            value.reset()
```

File: pytorch_lightning/trainer/trainer.py [hunk 5]
```diff
@@ -747,6 +744,9 @@ def run_evaluation(self, on_epoch=False):
 
         torch.set_grad_enabled(True)
 
+        # reset cached results
+        self.logger_connector.reset()
+
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:749]: nit: why not move this to right after L737?

**Review by ananthsub — APPROVED** [review]

**Comment by ananthsub** [comment]
thanks for the fix @SkafteNicki !

**Review by ethanwharris — APPROVED** [review]
Working for me :smiley:

**Comment by edenlightning** [comment]
Nice job @SkafteNicki, thanks for the swift fix.

File: pytorch_lightning/trainer/trainer.py (hunk 5 above)
Comment (SkafteNicki) [reply pytorch_lightning/trainer/trainer.py]: changed!

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 3 above)
Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:244]: I am not completely sure about what is going on myself, but apparently the `self._internal_type` is equal to `ResultStoreType.INSIDE_BATCH_TRAIN_LOOP` even when we are at epoch end for all training metrics. `self._internal_type` is equal to `ResultStoreType.OUTSIDE_BATCH_TRAIN_LOOP` for validation metrics (atleast when this `reset` function is called).

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 6]
```diff
@@ -233,6 +233,18 @@ def auto_reduce_results_on_epoch_end(self) -> None:
...
         self.has_reduced = True
 
+    def reset(self) -> None:
+        """
+        Call at the end of epoch to reset Result objects
+        """
+        for dl_idx in range(self.num_dataloaders):
+            epoch_metrics = self._internals[dl_idx] if not self.has_reduced else self._internals_reduced[dl_idx]
+            if self._internal_type == ResultStoreType.INSIDE_BATCH_TRAIN_LOOP:
+                for opt_idx in list(epoch_metrics):
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:243]: do you need the list casting here?
```suggestion
                for opt_idx in epoch_metrics:
```

**Review by Borda — APPROVED** [review]
lgtm

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 6 above)
Comment (SkafteNicki) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:243]: just reused the same logic from here:
https://github.com/PyTorchLightning/pytorch-lightning/blob/30b7440e123d5a91806b493995293e97a2ab80f8/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py#L195-L199
so not sure if needed

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:243]: It is not. In the snippet you shared, `epoch_metrics` gets modified inside the loop (with `del`). That's why we make the cast to the dict keys.

**Review by tchaton — APPROVED** [review]
LGTM !
